### PR TITLE
#168365057 Implementation of a reusable in-App notification module

### PR DESCRIPTION
**Pull request by Wokoro**
**What does this PR do?**

  Provides :+1: realtime notification for the application through dedicated routes.

  **Description of Task to be completed?**

  Implementation of a module that can be reused were realtime notification is required. 

The module use Server Sent Events(SSE) and sequelize hooks to accomplish the feature. Users who which to receive a given set of notification would be connected via a dedicated `get` route.

  **How should this be manually tested?**

  #### Step 1
Clone repository `clone git@github.com:andela/cyclops-backend.git`

  #### Step 2
install project packages `npm install`.

#### Step 3
Create the controller for the notification. An example that creates a controller notification for user `after creation` and `before creation` can be found [here](https://github.com/andela/cyclops-backend/blob/ch-inapp-notification-setup-168365057/src/controllers/ExampleUserNotificationController.js)
   
#### Step 4
Expose a routes which clients would receive notifications from. An example of an endpoint that would notify users of  new `user creation` can be found [here](https://github.com/andela/cyclops-backend/blob/ch-inapp-notification-setup-168365057/src/routes/notifications.js)

**Note :** All commands above should be ran in the project root directory.5

  **Any background context you want to provide?**

 * During initial connection developers are reponsible for sending client unread notifications by querying the notifications database table, after which events would be added for sending New notifications.

 * Developers are responsible for structuring the **message** to be sents and setting the type of **events** clients can listen for, message **id** to be sent etc. This allow wide variety of application of the module.

  *  In other for the front end to consume response data, `EventSource` API would be used.

  *  Data to be sent should have either a `\n`  or `\n\n` at the end. `\n` sends data while `\n\n` indicates the end of a message segment.

  *  Middlewares can be added just like in the case of normal routes. An example is given below

    // An example addition of authenticateUser middleware to the route.
    notificationRoutes.get(  '/request',  authenticateUser, notificationController.requestNotifications );`

  *  Multiple notifications can be triggered through one endpoint. An example of this is presented below

    // Route to exposed all user events Live notification endpoint for user creation.
    notificationRoutes.get('/user', userNotificationController.userNotifs)

    // Inside UserNotification Controller
      async create(req, res, next) {
        setHeaders(req, res);
        try {
          // Add realtime notification for user account creation.
            addHook(User, 'afterCreate', 'notification', (user) => {
            // code comes here
        });

        // Add realtime notification. for user update.
          addHook(User, 'afterUpdate', 'update', () => {
          //code comes here
        });
      } catch (error) {
        next(error);
      }
    }

* Token would be passed as query parameter `slt`,  for routes that need authentication.

* Hooks when used, should be spread across different models instead of a single model in this case `Notifications` model. This design structure helps to increase performance.

* Available notifications triggers.

1. Model update (before and after)

2. Model creation (before and after)

3. Model update (before and after)

**Note :** `Model` in the above statements refers to an instance of `sequelize model`

  **What are the relevant stories?**

  [#168365057](https://www.pivotaltracker.com/story/show/168365057)

  **Screenshots (if appropriate)**

    Example account creation response image
realtime

  **Questions:**

 none

### CODE REVIEW

**Comment by Wokoro** [comment]
Todo: 
* Reduce number of commits to one.
* Increase test coverage.

**Review by justinefe — COMMENTED** [review]
The edits are just too much and i'm not even seeing the reuseable in app - notification function maybe you could just point it out or something just my opinion though

File: src/routes/index.js [hunk 1]
```diff
@@ -1,18 +1,25 @@
...
+  app.get('/api/v1/', (req, res) => res.status(200).send({
     status: 'success',
     data: 'Welcome to the Cyclops Barefoot Nomad backend API'
   }));
 
   app.use('/api/v1', [user, trip, office]);
 
-  app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDoc));
+  // Add notification endpoints to application
+  app.use('/api/v1/notifications', notifications);
```
Comment (richienabuk) [inline src/routes/index.js:20]: Nice job. Seen a lot of test.
I'd add the route to the array as `app.use('/api/v1', [user, trip, office, notifications]);`
Is there any specific reason you separated it?

Comment (Wokoro) [reply src/routes/index.js:20]: No functionality attached, I didn't want to repeat the `notications` word when defining subsequent  routes that's all.

File: tests/userRole.test.js [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import sinon from 'sinon';
```
[the remark this answers is not in the document]
Comment (dieudonneAwa) [reply tests/userRole.test.js:1]: You PR looks good, well done. But please take care of this

**Review by dieudonneAwa — APPROVED** [review]
Please fix the Hound violation

File: tests/userRole.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Wokoro) [reply tests/userRole.test.js:1]: Ok. I would ASAP.

**Pull request merged**